### ISSUE TITLE
Mockups for Template workflow

### ISSUE THREAD

**Issue by hermwong** [phonegap/phonegap-app-desktop#569]
Parent task #543

**Comment by jahoni** [comment]
Put together a couple of mock-ups (no prototype to show animations, but maybe we can talk through that, @hermwong)

Find them here: [retina](https://www.dropbox.com/s/iml41xyjo4txbt1/mock-up.psd?dl=0) & [non-retina](https://www.dropbox.com/s/dx5xe2s1jxzq3gi/mock-up_nonretina.psd?dl=0) PSDs (layer comps named for states)

![desktop-app_0009_add-create-step1](https://cloud.githubusercontent.com/assets/178345/13681774/c265a33a-e6b3-11e5-90a6-7b01df9c83f8.png)

![desktop-app_0010_add-create-step2](https://cloud.githubusercontent.com/assets/178345/13681778/c7fc788c-e6b3-11e5-8386-a91c4cd1c067.png)

**Comment by chrisgriffith** [comment]
I like the design, it works for me. A few questions: Will the template list be scrollable? And where is this list pulled from? Can the end user configure it?

**Comment by hermwong** [comment]
The initial release of this feature will only have the 5 listed default templates. We're are planning to add in the ability for users to add/use their own templates in the future.

**Comment by chrisgriffith** [comment]
So what is the 5th template? I assume Snotify is placeholder. I would vote for Ionic. What are the thoughts on the initial ordering? Blank, Hello World, then the other 3?

**Comment by hermwong** [comment]
We decided on using Hello World as the default template since it's what the Desktop app currently uses in the Create New Project workflow.

**Comment by jahoni** [comment]
hey @chrisgriffith, thanks for your interest.

as @hermwong, at this point we are sticking with the 5 templates that the PhoneGap team has put together, and thus has the capacity to maintain. When more templates are available/searchable, we will like look at revising the interaction paradigm.

The order of templates is based on their likely usefulness to most users, and the initial release of this feature will see the templates ordered as shown in the mock-up. We'll be incorporating analytics to determine if the order could be revised, but given the low number I'm not very concerned about people "missing" what is available. 

Snotify is a placeholder for an example app the @devgeeks is working on. It does not use Ionic, but rather Framework7.

**Comment by chrisgriffith** [comment]
That sounds fine. I wonder if there should be two Framework 7 templates
shown? I know that we can't support every framework, but I think we should
support the main ones.

**Comment by mwbrooks** [comment]
**TLDR;** PhoneGap CLI supports installing any third-party template from npm. PhoneGap Desktop plans to do this in the second phase of the templates feature.

One point that hasn't been covered in this discussion is that Cordova templates can be installed from npm. Currently, that's how all of the above PhoneGap-maintained templates are installed.

Unfortunately, it's not easy to search for Cordova templates on npm and even more difficult to do it programmatically. All templates must use the keyword `cordova:template`. You can then discovery templates by [listing all packages with the keyword `cordova:template`](https://www.npmjs.com/browse/keyword/cordova:template). However, you cannot search for keywords. This same issue came up for Cordova plugins, which is why they host their own service to search for plugin (plugins.cordova.io). They plan to do the same for templates, but there is no deadline set.

The PhoneGap CLI just accepts that it's awkward. You can still install any template by specifying the npm package name in `phonegap create my-app --template my-package-name`. If you happen to specify an invalid package name (say `grunt`), then well... it'll still try to create a project from that package. You can also list recommendd PhoneGap templates using `phonegap template list` and type `phonegap template search` to open your browser to the npm page that displays all templates.

The PhoneGap Desktop decided that specifying third-party templates can come in the second release. The first release can support the PhoneGap-maintained templates (equivalent to `phonegap template list`). Discovery of third-party templates cannot happen until Cordova hosts a template search service (or someone else does). Installation of third-party templates can happen now, but without discovery it's a little sad and very error prone. This is why it's been pushed back to a later release.

The final option is what @chrisgriffith has been touching on. PhoneGap could include common third-party templates in the recommend list (shown on the PhoneGap Desktop and `phonegap template list`). This is totally reasonable and we'd love to do it! However, as of today, [PhoneGap is the only project to be publishing Cordova templates to npm](https://www.npmjs.com/browse/keyword/cordova:template).

Maybe that helps a little?

**Comment by chrisgriffith** [comment]
Yes, as an outsider, this is great information behind all of this. My basic
request is since PhoneGap is supporting a Framework7 template and a React
template, how can it 'officially' have an Ionic one in the mix? I am just
thinking that this list to reflect the popular frameworks to the user. But
that is just my personal preference.

Chris

**Comment by hermwong** [comment]
Outstanding items:

1. reset state of template list when:
  1. cancel out of create new project
  1. initiate create new project workflow
1. retain state of template list when:
  1. `back` pressed from project details overlay
1. `create project` pressed in project details overlay should create new project using selected template

**Comment by hermwong** [comment]
Had a brief discussion with @jahoni regarding the transitions between the Template Overlay and the Project Details Overlay.

Template Overlay

- Next clicked
  - slide out left
- Cancel clicked
  - slide out up

Project Details Overlay

- Create clicked
  - slide out up
- Back clicked
  - slide out right
- Cancel clicked
  - slide out up

When the `Next` button is clicked on the Template Overlay, the Project Details Overlay should slide in from the right simultaneous to the Template Overlay's slide out left animation.

**Comment by hermwong** [comment]
Preview of updated transitions.

![screen recording 2016-03-31 at 03 03 pm](https://cloud.githubusercontent.com/assets/107754/14192380/287f03f6-f752-11e5-9143-5439b0ec3ab6.gif)

**Comment by jahoni** [comment]
@hermwong it bothers me that the two step screens are not the same height - step 2 is about 20 pixels longer. I'd rather that it looks like a single card sliding, with the bottom buttons & bottom dark background aligned.

**Comment by hermwong** [comment]
Template overlay & project details overlay at same height.

We probably need to adjust the spacing of the elements on the template overlay.

![screen recording 2016-04-21 at 05 12 pm](https://cloud.githubusercontent.com/assets/107754/14727892/6b7ae088-07e4-11e6-912c-e26c6153f2fc.gif)

**Comment by jahoni** [comment]
hey @hermwong - yeah...ideally the buttons (and the dotted line above them) are at the same position on the two cards/screens.

**Comment by hermwong** [comment]
UI has been wired in.
